doc/compiler/osr/OSR.md

```diff
@@ -10,36 +10,241 @@ is available at https://www.apache.org/licenses/LICENSE-2.0.
+in the interpreter at the appropriate point.
 
-OSR transition is the process of transfering the execution from jitted code to the interpreter. OSR transition involves leaving the jitted code, reconstructing the runtime VM state and resuming the execution where the jitted code left off in the interpreter.
+# High Level Sequence of Events
 
-# OSR point
-An OSR point is a place in the jitted code where OSR transition is supported.
+OSR is typically triggered by some VM event that applies to a JITed method that is currently on stack.
+For example, such event can be setting a breakpoint in a method on stack.
+Changing some assumptions about JITed method that is currently on stack (invalidation) can eventually lead to OSR as well.
 
-At an OSR point, the VM state can be correctly reconstructed and is consistent with the VM state needed by the interpreter to resume the execution at a given bytecode index.
+A VM event can be detected at certain points during execution of a JITed code. These points are called **OSR yield points**. JIT and VM agree on the set of the OSR yield points.
 
-To achieve that, bookkeeping is required at all possible OSR points during ILGen. Due to the cost and complexity of bookkeeping, OSR transition is limited to some well-defined points, which are calls, asyncchecks and monitor enters, etc.
+Usually, we would like the VM event to be discovered relatively soon after it took place. Garbage collection (GC) points meet such criteria.
+Therefore, depending on the event, OSR yield points usually include all or a subset of GC points.
+The reason why we would like to limit the set of the OSR yield points, if possible, is the cost of the bookkeeping associated with each such point as well as the restrictions that they impose on the optimizer.
 
-Being part of the program, OSR points like calls, asyncchecks and monitor enters can be removed or optimized and no longer be identified as OSR points. However, OSR transition may still be supported at that position.
+The actual OSR happens at **OSR Induction points**. They can happen at the time of the yield or later, as descitbed in the following sections.
 
-# potentialOSRPointHelper
-potentialOSRPointHelper is a call that serves as a marker of places where OSR transition is supported. The target bytecode index to be executed in the interpreter can be obtained by adding an offset (an information stored in the call node, the size of OSR point's bytecode in post-execution OSR) to the bytecode index of the helper call node.
+Therefore, a high level sequience of events can be described as following:
 
-The helper is not a must for OSR, but it can help locating the OSR points that are available at ILGen but disappear due to later transformations. It works as a bookkeeping tool, to track places that are OSR possible.
+    VM event => OSR yield => OSR induction => OSR transition
```

Comment:
I thought the given description is detailed enough and already adds considerably more details and clarity to the old one, especially from the JIT prospective. But we can expand even more...

Reply:
Changed to the following:

_Note that sometimes the VM event does not finish until OSR transition is done_. Therefore, _in the most general case,_ a high level sequence of events can be described as following:

    `VM event initiation => OSR yield => OSR induction => OSR transition => VM event completion`

Reply:
Sorry, I think I've led you to make this a bit worse than it was before. If we consider the class redefinition example again with next-gen HCR, then an affected compiled body could be running at the time of redefinition, and it could potentially continue running indefinitely after the event completion without ever inducing OSR

I think the core of what I was trying to get at is that the OSR yield synchronizes all threads with the VM. Maybe it would be better to describe the VM event and OSR yield as happening simultaneously? e.g.
```
(VM event + OSR yield) => ...
```
Otherwise, I think your original sequence still makes sense if we think of the VM event as being the point at which the VM requests that threads yield. It would just be missing what IMO is a crucial detail, that being how the VM waits for all threads to yield before actually violating any assumptions

Reply:
I would divide it into two issues:

1. The fact that "an affected compiled body could be running at the time of redefinition, and it could potentially continue running indefinitely after the event completion without ever inducing OSR is addressed in the openj9 document describing OSR-based HCR
2. What exactly is a VM event. I was thinking of it on a high level, from a user point of view: setting a breakpoint , redefining a method, or changing the value of a static final. That's why I wanted to separate it from the yield point. Basically, user makes some request, but the VM/JIT might react to it later, at some agreed points. The VM event is described int the first paragraph of this section. Maybe we can modify it accordingly? Or maybe we can introduce "User request" ?

Reply:
> 1\. The fact that "an affected compiled body could be running at the time of redefinition, and it could potentially continue running indefinitely after the event completion without ever inducing OSR is addressed in the openj9 document describing OSR-based HCR

Yes, I don't think that necessarily needs to be directly addressed here. It's just that it's now in contradiction to the updated sequence, which says that the VM event completes after OSR transition

> 2\. What exactly is a VM event. [...] setting a breakpoint, redefining a method, or changing the value of a static final.

It's hard to say that these really take place before the OSR yield. Until all threads have stopped, the program is still running, and during that time the breakpoint is not yet set (i.e. we can still run that code without hitting it), the class has not yet been redefined (i.e. we can still enter its original methods), the static final has not yet been modified (i.e. we can still load it and see its original value)

If we expand our notion of the VM event to encompass both the initiation/request/attempt and its actual effect, then now the OSR yield happens during the VM event, which is what I was trying to describe above: initiation ⇒ OSR yield ⇒ completion. But splitting up the VM event this way complicates the description because the completion and the OSR induction are not ordered w.r.t. each other

This is related to the assumption invalidation vs. violation distinction that I suggested elsewhere

Thinking on this some more, it seems my real (original) issue here is that if we say that the VM event comes first and the OSR yield later, and that a VM event is e.g. redefining a class, then to me that sounds like it says that we can redefine a class (i.e. finish redefining it) with the compiled code somehow only needing to yield after the fact

*Request* might be good terminology. Personally I'd try to avoid *user*, since I don't think it's clear who the user is supposed to be. Maybe *VM request*, which would be when the VM asks threads to stop as part of one of the high-level events?

Reply:
I agree that _request_ describes better (and on enough level of detail) what I was trying to say originally. What about going to the old version and changing _event_ to _request_? So it will be like this:

OSR is typically triggered by some VM request that applies to a compiled method that is currently on stack.
For example, such request can be setting a breakpoint in a method on stack.
In addition, if assumptions that the compiler made while compiling a method are invalidated, OSR might eventually be triggered for the method as well.


A VM request can be detected at certain points during execution of compiled code. These points are called **OSR yield points**. The JIT and the VM agree on the set of the OSR yield points.

Usually, we would like the VM request to be discovered relatively soon after it took place. [Garbage Collection (GC) points](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298) meet such criteria.
Therefore, depending on the request, OSR yield points usually include all or a subset of GC points.
The reason why we would like to limit the set of the OSR yield points, if possible, is the cost of the bookkeeping associated with each such point as well as the restrictions that they impose on the optimizer.

The actual OSR happens at **OSR Induction points**. They can happen at the time of the yield or later, as described in the following sections.

High level sequence of events can be described as follows:

    VM request => OSR yield => OSR induction => OSR transition

Reply:
I feel like *event* is still a better description except in this one instance. How about just this?
> *As part of such an event, the VM requests that threads stop, which* can be detected [...]
> [...]
> ```
> VM request => OSR yield => OSR induction => OSR transition
> ```

If you still prefer to go with *request* throughout, then I'd suggest saying that setting a breakpoint is an example of something that can cause a request, rather than that it is itself an example of a request

Reply:
Gita and I agreed that the original sequence (starting with *VM event*) makes sense if we describe the VM event as an *attempt.* So rather than the event being the setting of a breakpoint, it's an *attempt* to set a breakpoint